control-plane/agents/src/bin/core/volume/operations.rs

```diff
@@ -980,49 +888,60 @@ impl CreateVolumeExe for CreateVolume {
+        candidates: Self::Candidates,
     ) -> Vec<Replica> {
-        let mut replicas = Vec::<Replica>::with_capacity(candidates.candidates().len());
-        for replica in candidates.candidates() {
-            if replicas.len() >= self.replicas as usize {
-                break;
-            } else if replicas.iter().any(|r| {
-                r.node == replica.node
-                    || spread_label_is_same(r, replica, context).unwrap_or_else(|error| {
+        let mut fina_replicas = Vec::<Replica>::new();
+        for crc in candidates.iter() {
+            let mut replicas = Vec::<Replica>::with_capacity(crc.candidates().len());
+            for replica in crc.candidates() {
+                if replicas.len() >= self.replicas as usize {
+                    break;
+                } else if replicas.iter().any(|r| {
+                    r.node == replica.node
+                        || spread_label_is_same(r, replica, context).unwrap_or_else(|error| {
+                            context.volume.error(&format!(
+                                "Failed to create replica {:?} for volume, error: {}",
+                                replica,
+                                error.full_string()
+                            ));
+                            false
+                        })
+                }) {
+                    // don't re-use the same node or same exclusion labels
+                    continue;
+                }
+                let replica = if replicas.is_empty() {
+                    let mut replica = replica.clone();
+                    // the local replica needs to be connected via "bdev:///"
+                    replica.share = Protocol::None;
+                    replica
+                } else {
+                    replica.clone()
+                };
+                match OperationGuardArc::<ReplicaSpec>::create(context.registry, &replica).await {
+                    Ok(replica) => {
+                        replicas.push(replica);
+                    }
+                    Err(error) => {
                         context.volume.error(&format!(
                             "Failed to create replica {:?} for volume, error: {}",
                             replica,
                             error.full_string()
                         ));
-                        false
-                    })
-            }) {
-                // don't re-use the same node or same exclusion labels
-                continue;
+                        // continue trying...
+                    }
+                };
             }
-            let replica = if replicas.is_empty() {
-                let mut replica = replica.clone();
-                // the local replica needs to be connected via "bdev:///"
-                replica.share = Protocol::None;
-                replica
+
+            if replicas.len() < context.volume.as_ref().num_replicas as usize {
+                self.undo(context, replicas).await;
```

Comment:
Why don't we handle the errors here?

Reply:
because we want to try out the set of other replica candiates. 
If all the elements in the set of replica candiates are over then we error out , which is done in the `run` method